### PR TITLE
ARROW-2724: [Packaging] Determine whether all the expected artifacts are uploaded

### PR DESCRIPTION
**Pull request by kszucs**
Also resolves [ARROW-2718: [Packaging] GPG sign downloaded artifacts](https://issues.apache.org/jira/browse/ARROW-2718)

Contains a sample configuration for running nightly builds.

### CODE REVIEW

**Comment by kszucs** [comment]
@xhochy Orc changes broke `conda-osx` and `conda-linux` https://travis-ci.org/kszucs/crossbow/jobs/396585152. I guess this commit https://github.com/apache/arrow/commit/4db8a369d3df3af43d5023db1b376ece8b17cd54

File: dev/tasks/tasks.yml [hunk 1]
```diff
@@ -17,38 +17,73 @@
...
     platform: linux
     template: conda-recipes/travis.linux.yml
+    artifacts:
+      - arrow-cpp-{version}-py27_0-linux-64.tar.bz2
+      - arrow-cpp-{version}-py35_0-linux-64.tar.bz2
+      - arrow-cpp-{version}-py36_0-linux-64.tar.bz2
+      - pyarrow-{version}-py27_0-linux-64.tar.bz2
+      - pyarrow-{version}-py35_0-linux-64.tar.bz2
+      - pyarrow-{version}-py36_0-linux-64.tar.bz2
+      - parquet-cpp-1.4.0-0-linux-64.tar.bz2
```
[the remark this answers is not in the document]
Comment (kszucs) [reply dev/tasks/tasks.yml]: I'm not sure that we want to build parquet-cpp here too (in the upstream parquet-cpp conda package arrow's version is pinned currently).

We can of course incorporate parquet version somehow.

File: dev/tasks/conda-recipes/arrow-cpp/bld.bat [hunk 2]
```diff
@@ -21,6 +21,7 @@ pushd "%SRC_DIR%"\cpp\build
 set ARROW_BUILD_TOOLCHAIN=%LIBRARY_PREFIX%
 
 cmake -G "%CMAKE_GENERATOR%" ^
+      -DARROW_VERBOSE_THIRDPARTY_BUILD=ON ^
```
[the remark this answers is not in the document]
Comment (kszucs) [reply dev/tasks/conda-recipes/arrow-cpp/bld.bat]: For now I'll turn it off, later we can make the templating more flexible.

File: dev/tasks/crossbow.py [hunk 3]
```diff
@@ -391,45 +422,101 @@ def submit(task_names, job_prefix, config_path, dry_run, arrow_path,
...
+
+            msg = click.style(tpl.format('SIGNED'), fg=COLORS['ok'])
+            click.echo(msg + artifact)
+
+            signatures.append(signature_path)
+
+    # upload the signatures
+    if upload:
+        queue.upload_assets(job, signatures, token=github_token,
+                            content_type='application/pgp-signature')
```
Comment (wesm) [inline dev/tasks/crossbow.py]: What's the thinking re: uploading the GPG signatures? We only need the sigs right before uploading the artifacts to the ASF's dist system right before a release vote

Comment (kszucs) [reply dev/tasks/crossbow.py]: I'm not exactly aware of the steps, I thought the release vote uses the assets uploaded to github, then comes the ASF's dist system (which is what?), bintray, forge, etc.

Comment (kszucs) [reply dev/tasks/crossbow.py]: BTW here are the artifacts and signatures: https://github.com/kszucs/crossbow/releases

Should I skip uploading the signatures?

Comment (xhochy) [reply dev/tasks/crossbow.py]: Uploading signatures in the script is not necessary. A release manager should compute the signatures on all artifacts locally before they are committed to the ASF svn for distribution.

**Comment by kszucs** [comment]
@wesm @xhochy may I ask a review?

I've removed the signature upload.
After https://github.com/apache/arrow/pull/2169 conda and wheel builds are working again.

crossbow_status

Linux packages are still failing but I'm continuously apply kou's patches in order to fix them.

**Comment by wesm** [comment]
Yes, will take a closer look today. The status output is nice! =)

**Comment by kszucs** [comment]
At least highlights when something is missing.

**Comment by kszucs** [comment]
@wesm @cpcloud 

The linux build are working now, thanks to @kou!
Here You can see linux artifacts too: https://github.com/kszucs/crossbow/releases

However debian packages have [naming conflicts](https://github.com/apache/arrow/pull/2162/files#diff-0abe5ae3589de012350559065a4194a6R91), ideas how should I resolve them? Postfixes?

**Comment by xhochy** [comment]
> However debian packages have naming conflicts, ideas how should I resolve them? Postfixes?

Probably prefix them, this keeps the file extension as-is which makes them easier to handle. @kou Any better ideas? 

We could also encode the distribution in the version but that is actually something distributions only do when they have applied patches themselves.

**Comment by kszucs** [comment]
I was thinking of libarrow0-dbgsym_{version}-1_amd64_{distro}.deb` instead of `libarrow0-dbgsym_{version}-1_amd64.deb`

**Comment by xhochy** [comment]
Fine with me

**Comment by kou** [comment]
How about creating `debian-stretch-{version}.tar.gz` that includes all `.deb` files for Debian stretch?
We need to download all `.deb` and `.rpm` to test these files. It's useful that we can download all related files.

**Comment by kszucs** [comment]
@kou Uploading tarballs would involve more changes, so I've created a [jira task](https://issues.apache.org/jira/browse/ARROW-2792) to track it. 

@cpcloud I suggest to close this one and resolve the questionable parts via subsequent PRs. After merging We can build [nightlies](https://travis-ci.org/kszucs/crossbow/builds/399895987) against arrow's master instead of [the current PR](https://github.com/kszucs/crossbow/blob/master/.travis.yml#L45)

**Comment by kou** [comment]
We can work on more packaging tasks such as the followings based on this work.

  * https://issues.apache.org/jira/browse/ARROW-2383: [C++] Debian packages need to depend on libprotobuf
  * https://issues.apache.org/jira/browse/ARROW-2791: [Packaging] Build Ubuntu 18.04 packages
  * https://issues.apache.org/jira/browse/ARROW-2792: [Packaging] Consider uploading tarballs to avoid naming conflicts

So I'll merge this pull request in a few days if nobody objects.

**Review by wesm — COMMENTED** [review]
Some minor comments / questions, but this is great to see, will be anxious to see this move toward something we can use to produce the 0.10.0 release

Some questions:

* How would building from a release tag work? and relatedly:

* Need to write a release management guide that includes using these tools. I'd like to move the release management documentation to the wiki out of GitHub https://cwiki.apache.org/confluence/display/ARROW/Apache+Arrow+Home

**Comment by wesm** [comment]
I started moving our release management guide to Confluence: https://cwiki.apache.org/confluence/display/ARROW/Release+Management+Guide

**Comment by kou** [comment]
I've merged the @wesm's work.
Now, updating our release management guide is out of scope of this pull request.

File: dev/tasks/conda-recipes/appveyor.yml [hunk 4]
```diff
@@ -44,8 +44,8 @@ install:
 build: off
 
 test_script:
-  - git clone -b {{ ARROW_BRANCH }} {{ ARROW_REPO }} arrow || exit /B
-  - git -C arrow checkout {{ ARROW_SHA }} || exit /B
+  - git clone -b {{ arrow.branch }} {{ arrow.remote }} arrow || exit /B
+  - git -C arrow checkout {{ arrow.head }} || exit /B
   - pushd arrow\dev\tasks\conda-recipes
   - conda.exe build --output-folder . parquet-cpp arrow-cpp pyarrow
```
[the remark this answers is not in the document]
Comment (kszucs) [reply dev/tasks/conda-recipes/appveyor.yml]: Conda builds them in topological order, see appveyor's [log](https://ci.appveyor.com/project/kszucs/crossbow/build/1.0.259/job/1aymabx7vkfbywbp?fullLog=true)

File: dev/tasks/conda-recipes/appveyor.yml [hunk 5]
```diff
@@ -70,4 +70,4 @@ deploy:
 notifications:
   - provider: Email
     to:
-      - {{ EMAIL }}
+      - {{ job.email }}
```
[the remark this answers is not in the document]
Comment (kszucs) [reply dev/tasks/conda-recipes/appveyor.yml]: No, I've received tons of success, error, failure messages. We can tune it later though.

**Comment by kszucs** [comment]
Sorry, I've missed your review. 

- ARROW-2713 [Packaging] Fix linux package builds
- ARROW-2770 [Packaging] Account for conda-forge compiler migration in conda recipes
- ARROW-2792 [Packaging] Consider uploading tarballs to avoid naming conflicts (I need comments on this one)
- ARROW-2713 [Packaging] Fix linux package builds (this PR resolves it)
- ARROW-2383 [C++] Debian packages need to depend on libprotobuf (is this about [adding](https://github.com/kszucs/arrow/blob/454040643e81c233e53f7cceb45b4df1b71430ca/dev/tasks/linux-packages/Rakefile#L86) `ARROW_PROTOBUF_USE_SHARED`?)

**Comment by kszucs** [comment]
I've incorporated the recent conda-forge changes. Added build [variants](https://github.com/kszucs/arrow/tree/eb06a0d0e051ecaa363b0c965d5a75999194d449/dev/tasks/conda-recipes/variants) too, which changes the [build hashes](https://github.com/kszucs/crossbow/releases/tag/build-183).

File: dev/tasks/crossbow.py [hunk 6]
```diff
@@ -64,9 +68,10 @@ def credentials(self, url, username_from_url, allowed_types):
 
 class Repo(object):
```
[the remark this answers is not in the document]
Comment (kszucs) [reply dev/tasks/crossbow.py]: Not too verbose one, but added.

**Comment by wesm** [comment]
Hm, maybe the Windows issue is transient. I have been seeing the hashes in conda package names after the compiler migration. I also don't know whether it's a problem.

Shall we merge this and move on to the next stage of work? Maybe let's send an update to the mailing list to summarize where things stand viz-a-viz creating the 0.10.0 release

**Comment by kszucs** [comment]
I've tried to run it three times, sadly with the same outcome. 

@wesm Merge it, I'll do the rest in followup PRs. I can send an email tomorrow, before the sync call.

**Comment by wesm** [comment]
No problem. +1 -- the Travis CI failure is due to launchpad.net flakiness

**Pull request closed without merging**